include/ppx/geometry.h

```diff
@@ -169,68 +169,64 @@ class Geometry
         uint32_t    GetElementCount() const;
         uint32_t    GetDataSize() const;
 
-        // Append() will automatically adjust mOffset, but it may need to be
-        // manually adjusted if Overwrite() is used
-        void SetOffset(uint32_t offset) { mOffset = offset; }
-
         // Trusts that calling code is well behaved :)
         //
+        // Write(count, pValues...) is used to write a chunk of data at specified offset
+        // If called with offset=-1, will write at the end and adjust mOffset accordingly
         template <typename T>
-        void Append(const T& value)
-        {
-            uint32_t sizeOfValue = static_cast<uint32_t>(sizeof(T));
-
-            if (mMaxElementCount == 0) {
-                // Allocate storage for incoming data
-                SetSize(mOffset + sizeOfValue);
-            }
-
-            // Copy data
-            const void* pSrc = &value;
-            void*       pDst = mData.data() + mOffset;
-            memcpy(pDst, pSrc, sizeOfValue);
-
-            mOffset += sizeOfValue;
-        }
-
-        template <typename T>
-        void Append(uint32_t count, const T* pValues)
+        void Write(uint32_t count, const T* pValues, int32_t offset = -1)
         {
             uint32_t sizeOfValues = count * static_cast<uint32_t>(sizeof(T));
+            uint32_t writeOffset  = offset == -1 ? mOffset : offset;
 
-            if (mMaxElementCount == 0) {
-                // Allocate storage for incoming data
-                SetSize(mOffset + sizeOfValues);
+            if (offset == -1 && !mFinalSizeSet) {
+                // New data is being appended, need to allocate storage
+                SetSize(writeOffset + sizeOfValues);
+            }
+            else {
+                PPX_ASSERT_MSG(writeOffset + sizeOfValues <= GetSize(), "Attempting to write values of size " << sizeOfValues << " at offset " << offset << " will overflow the buffer of size " << GetSize());
             }
 
             // Copy data
             const void* pSrc = pValues;
-            void*       pDst = mData.data() + mOffset;
+            void*       pDst = mData.data() + writeOffset;
             memcpy(pDst, pSrc, sizeOfValues);
 
-            mOffset += sizeOfValues;
+            if (offset == -1) {
+                mOffset += sizeOfValues;
+            }
         }
 
-        // If Overwrite is used to write new data, calling code will also
-        // have to SetOffset() appropriately
-        //
+        // Write() is used to write data at the specified offset
+        // If called with offset=-1, will write at the end and adjust mOffset accordingly
         template <typename T>
-        void Overwrite(const T& value, size_t offset)
+        void Write(const T& value, int32_t offset = -1)
```

Comment:
Can this function call Write(1, &value, offset) instead of duplicating the code?